bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py

```diff
@@ -0,0 +1,512 @@
+        operation.args[1] = path_with_range
+
+        # Store Snapshot result index -> path
+        self.snapshots.append(
+            CppWorkflow.SnapshotData(self.res_ptr_id, path_with_range))
+
+    def _handle_asnumpy(self, operation, parent_id):
+        '''
+        Since AsNumpy is a Python-only action, it can't be included in the
+        C++ workflow built by this class. Therefore, this function takes care
+        of saving the RDF node, generated in C++, on which an AsNumpy action
+        should be applied from Python.
+        This function also stores the index of the AsNumpy result in the final
+        list of results, where the index depends on the DFS traversal of the
+        computation graph.
+
+        Args:
+            operation (Operation): object representing the AsNumpy operation
+            
+            parent_id (int): id of the parent node in this workflow. Used to
+                save that node in the vector of output nodes.
+        '''
+
+        self.py_actions.append(CppWorkflow.PyActionData(self.res_ptr_id, operation))
+        self.graph_nodes += \
+            '\n  output_nodes.push_back(ROOT::RDF::AsRNode(rdf{}));' \
+            .format(parent_id)
+        self.res_ptr_id += 1
+
+    def _get_template_args(self, operation):
+        '''
+        Gets the template arguments with which to generate a given operation.
+
+        Args:
+            operation (Operation): object representing the operation whose
+                template arguments need to be returned.
+
+        Returns:
+            string: template arguments for this operation.
+        '''
+
+        # TODO: generate templated operations when possible, e.g. Max<double>
+
+        return ''
+
+    def _process_args(self, operation):
+        '''
+        Processes the arguments of an operation node.
+
+        Args:
+            operation (Operation): object representing the operation whose
+                arguments need to be processed.
+        '''
+
+        # TODO
+        # - Do a more thorough type conversion
+        # - Use RDF helper functions to convert jitted strings to lambdas
+
+        # Argument type conversion
+        for narg, arg in enumerate(operation.args):
+            if (narg > 0):
+                self.graph_nodes += ', '
+
+            if isinstance(arg, str):
+                self.graph_nodes += '"{}"'.format(arg)
+            elif isinstance(arg, tuple):
+                self.graph_nodes += '{'
+                for nelem, elem in enumerate(arg):
+                    if nelem > 0:
+                        self.graph_nodes += ','
+                    if isinstance(elem, str):
+                        self.graph_nodes += '"{}"'.format(elem)
+                    else:
+                        self.graph_nodes += '{}'.format(elem)
+                self.graph_nodes += '}'
+
+        # Make Snapshot lazy
+        # TODO: Do a proper processing of the args (user might have specified
+        # her own options object)
```

Comment:
Yes the old logic took care of this, still I would just leave this as it is for now. Handling this properly here requires getting the C++ pointer from the Python proxy and use that address in the C++ workflow (I think I can do this with `AddressOf` but I need to try). This is true not only for Snapshot options, but for any other object that RDF calls receive.

Anyway, merging this feature does not change anything unless you activate the `optimized` option. I'm more interested in doing a performance comparison of a first version of this to see if it's even worth investing more time in perfecting it.

Reply:
i totally agree on making the performance comparison and verifying that this is worth pursuing before fixing this -- but both things can happen before merging the feature :D 

(or we merge a feature with a known bug, but what's the benefit?)

Reply:
Since it's experimental, I don't see any problem in merging something that we know does not work for all cases (and it's not even enabled by default).

Just for the sake of (finer) granularity of the PR review process, I would like to have this in, test it, see where we are in terms of performance, then implement the compiled lambdas on top, another PR, review it, merge it, repeat performance tests. And so on and so forth for any later optimization we might think of.

And if at any stage we decide it's not worth pursuing anymore, we just kill it, it's experimental.

Reply:
we don't need the feature in upstream master in order to do testing and performance benchmarking.
the experimental namespace means "hey users, be warned that this API might change in the future", not "let's see what happens if we merge this, we can revert later if we change our mind".

with that said if you prefer merging I would suggest to at least print a warning if Snapshot is called with an options parameter and `distributed.optimized == True`.

Reply:
I do think that experimental also means we can change our mind. And like I said, this won't even be visible to users since its activation depends on a (now opaque) option.

Anyway, if we did what you suggest, we'd better move these PRs to a branch in my repo, instead of doing them to ROOT master. That would also allow the step-by-step review+merge+test cycles that I described.

I don't think that's necessary though, I'll add the warning.